### PR TITLE
fix(data): index themes in ES and use datafiller data

### PR DESCRIPTION
**Pull request by revolunet**
This PR contain several features/tradeoffs :

- The source of themes list in the UI now comes from ES rather than being built statically.
- The themes data are provided from datafiller API and ingested into ES like other data sources.
- The themes data also contain "selected results". Those are resolved "at runtime" on the API side. poorly implemented ATM but will be addressed in next PR.

I had to modify :

- fiches SP/MT theme resolution : theme resolution is done based on datafiller/themes.data.json 
- various UI stuff (breadcrumbs adjustements)

Todo :
 - testing
 - unify breadcrumbs in pages `themes` and `fiches`.
 - ~~fix fiches `parseReferences` : Digging into this, i also noticed an issue related to code-du-travail data structure and fiches resolution of "references juridiques". Will be addressed in another PR.~~

The datafiller now has a basic auth so you need to set env var `DATAFILLER_URL="https://user:pass@datafiller.num.social.gouv.fr" yarn start`

### CODE REVIEW

File: packages/code-du-travail-api/src/server/routes/search/search.savedResults.js [hunk 1]
```diff
@@ -1,4 +1,4 @@
-const knownQueries = require("@cdt/data...datafiller/prequalified.json");
+const knownQueries = require("@cdt/data...datafiller/prequalified.data.json");
```
Comment (lionelB) [inline packages/code-du-travail-api/src/server/routes/search/search.savedResults.js:1]: good idea to prefix our json data with `.data.json`

File: packages/code-du-travail-api/src/server/routes/themes/index.js [hunk 2]
```diff
@@ -1,15 +1,29 @@
...
 const elasticsearchClient = require("../../conf/elasticsearch.js");
 const getSearchByThemeBody = require("../search/searchByTheme.elastic");
+const {
+  getRootThemesQuery,
+  getThemeQuery,
+  getBySlug
+} = require("./search.elastic.js");
+
+// mapping elastic search source type -> route name
+const routeBySource = {
```
Comment (lionelB) [inline packages/code-du-travail-api/src/server/routes/themes/index.js]: I would recommand to make a packages with that so we can share it between data / api / frontend
using require("@cdt/data/sources") 
By the way if you find a better name than sources ;)

File: packages/code-du-travail-data/dataset/datafiller/package.json [hunk 3]
```diff
@@ -5,10 +5,12 @@
   "main": "index.js",
```
Comment (lionelB) [inline packages/code-du-travail-data/dataset/datafiller/package.json]: you should update this filed too

File: packages/code-du-travail-frontend/src/home/Themes.js [hunk 4]
```diff
@@ -37,15 +33,15 @@ export default class Themes extends React.Component {
     return (
       <Section>
         <Container>
-          {title && <Title>{title}</Title>}
+          <Title>{title || "Thèmes"}</Title>
```
Comment (lionelB) [inline packages/code-du-travail-frontend/src/home/Themes.js]: why did you remove the defaultProps ?

File: packages/code-du-travail-api/src/server/routes/search/search.savedResults.js (hunk 1 above)
Comment (revolunet) [reply packages/code-du-travail-api/src/server/routes/search/search.savedResults.js:1]: yep i just copied what you've done elsewhere :)

File: packages/code-du-travail-api/src/server/routes/themes/index.js [hunk 5]
```diff
@@ -36,49 +53,58 @@ router.get("/themes", ctx => {
...
-    ctx.throw(204, `there is no documents associated to the theme ${slug}`);
-  }
-  ctx.body = {
-    ...theme,
-    children: getChildren(slug),
-    breadcrumbs: getBreadcrumbs(slug),
-    documents: hits.hits
+  // rewrite refs to match documents inside ES index
+  // todo(perf): we should group ES resolution
+  const responseBody = {
```
Comment (lionelB) [inline packages/code-du-travail-api/src/server/routes/themes/index.js]: why did you update the api results format ? I think it would limit the changes on the frontend

File: packages/code-du-travail-api/src/server/routes/themes/search.elastic.js [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+function getBySlug({ source, slug }) {
+  return {
+    //_source: ["title", "slug", "position", "parents", "children", "refs"],
+    //sort: [{ position: { order: "asc" } }],
+    query: {
+      bool: {
+        must: [
+          {
+            term: {
```
Comment (lionelB) [inline packages/code-du-travail-api/src/server/routes/themes/search.elastic.js]: you can use filter since source is a keyword

File: packages/code-du-travail-data/dataset/datafiller/elastic.js [hunk 7]
```diff
@@ -20,7 +21,7 @@ const getReference = (result, ref) => {
     match: {
       slug: {
         query: slug,
-        minimum_should_match: "70%"
+        minimum_should_match: "95%"
```
Comment (lionelB) [inline packages/code-du-travail-data/dataset/datafiller/elastic.js]: does adding the title help or slug is enought to retrieve the items ?

File: packages/code-du-travail-data/dataset/datafiller/fetch-requetes.js [hunk 8]
```diff
@@ -1,21 +1,49 @@
...
+ fetch raw datafiller requetes data, filter and sort properly
 */
 
 const DATAFILLER_URL =
-  "https://datafiller.num.social.gouv.fr/kinto/v1/buckets/datasets/collections/requetes/records";
+  process.env.DATAFILLER_URL || "https://datafiller.num.social.gouv.fr";
+
+const RECORDS_URL = `${DATAFILLER_URL}/kinto/v1/buckets/datasets/collections/requetes/records`;
+
+// convert datafiller references to valid ES references
```
Comment (lionelB) [inline packages/code-du-travail-data/dataset/datafiller/fetch-requetes.js]: I would help to have a comment with the object's shape from kinto

File: packages/code-du-travail-api/src/server/routes/themes/index.js [hunk 9]
```diff
@@ -1,15 +1,29 @@
 const Router = require("koa-router");
 const API_BASE_URL = require("../v1.prefix");
-const {
-  getTheme,
-  getBreadcrumbs,
-  getChildren
-} = require("@cdt/data...themes/query");
 const elasticsearchClient = require("../../conf/elasticsearch.js");
 const getSearchByThemeBody = require("../search/searchByTheme.elastic");
```
Comment (lionelB) [inline packages/code-du-travail-api/src/server/routes/themes/index.js]: not used anymore :'(

File: packages/code-du-travail-data/dataset/datafiller/fetch-requetes.js [hunk 10]
```diff
@@ -1,21 +1,49 @@
...
 const sourcesPriority = [
   "faq",
   "fiche-service-public",
   "fiche-ministere-travail",
+  "external",
   "code-du-travail"
 ];
 
+const sortByKey = key => (a, b) => {
+  if (a[key] < b[key]) {
```
Comment (lionelB) [inline packages/code-du-travail-data/dataset/datafiller/fetch-requetes.js]: if a and b are string, return a[key].localeCompare(b[key])

File: packages/code-du-travail-data/dataset/datafiller/fetch-themes.js [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+*/
+
+const DATAFILLER_URL =
+  process.env.DATAFILLER_URL || "https://datafiller.num.social.gouv.fr";
+
+const RECORDS_URL = `${DATAFILLER_URL}/kinto/v1/buckets/datasets/collections/themes/records?_limit=1000`;
+
+const hasUrl = row => !!row.url;
+
+const getVariants = row => {
```
Comment (lionelB) [inline packages/code-du-travail-data/dataset/datafiller/fetch-themes.js]: this one is a duplicate of the fetch-requete.js. 
why not add a datafiller.utils.js to share common stuff

File: packages/code-du-travail-data/dataset/datafiller/fetch-themes.js [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+  } else if (a[key] > b[key]) {
+    return 1;
+  }
+  return 0;
+};
+
+const getChildren = (rows, row) =>
+  rows
+    .filter(node => node.parent === row.id)
+    .map(node => ({ ...node, slug: getSlug(node) }));
```
Comment (lionelB) [inline packages/code-du-travail-data/dataset/datafiller/fetch-themes.js]: getChildren could remove the refs prop of children theme (it avoid duplicate refs)

File: packages/code-du-travail-data/dataset/datafiller/fix-urls.js [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+};
+
+// check refs against ES
+const getReferences = row =>
+  Promise.all(
+    row.refs.filter(ref => ref.url).map(ref => replaceReference(row.title, ref))
+  ).then(arr => arr.filter(Boolean).sort(sortByKey("relevance")));
+
+const fixAll = async () =>
+  fetch(RECORDS_URL)
```
Comment (lionelB) [inline packages/code-du-travail-data/dataset/datafiller/fix-urls.js]: why not await (almost) all the things ?

File: packages/code-du-travail-data/dataset/datafiller/fix-urls.js [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+            ...row,
+            refs: await getReferences(row)
+          }))
+        )
+      )
+    );
+// DEBUG
+//.then(rows => console.log(JSON.stringify(rows, null, 2)) || rows);
+
+if (require.main === module) {
```
Comment (lionelB) [inline packages/code-du-travail-data/dataset/datafiller/fix-urls.js]: I wonder If we could have an updateDatafillerRef() that we can call in the fetch-requete that would limit the use of fix-url script

File: packages/code-du-travail-data/dataset/fiches_ministere_travail/report.js [hunk 15]
```diff
@@ -1,6 +1,6 @@
 const fiches = require("./fiches-min-travail.json");
 
-const hasNoTheme = fiche => !fiche.themeCdtn;
+const hasNoTheme = fiche => !fiche.parents || fiche.parents.length === 0;
```
Comment (lionelB) [inline packages/code-du-travail-data/dataset/fiches_ministere_travail/report.js]: I would prefer a themes key, to me, parent implies a "fiche" could have children

File: packages/code-du-travail-data/dataset/fiches_ministere_travail/updateTheme.js [hunk 16]
```diff
@@ -1,20 +1,33 @@
...
-    fiche.themeCdtn = themeMapping[slug];
-    if (!fiche.themeCdtn) {
-      console.error(`${fiche.title} - (${fiche.internalId}) has no theme`);
-      fiche.themeCdtn = [];
-    }
+    const ficheUrl = `/fiche-ministere-travail/${fiche.slug}`;
+    const theme = themes.find(theme =>
+      theme.refs.map(r => r.url.split("#")[0]).includes(ficheUrl)
+    );
+    fiche.theme = theme && theme.slug;
```
Comment (lionelB) [inline packages/code-du-travail-data/dataset/fiches_ministere_travail/updateTheme.js]: if no theme > warn and return also theme could be name themeSlug

File: packages/code-du-travail-data/dataset/fiches_service_public/package.json [hunk 17]
```diff
@@ -7,6 +7,7 @@
     "url": "https://github.com/SocialGouv/code-du-travail-numerique/issues"
   },
   "dependencies": {
+    "@socialgouv/legi-data": "^1.1.0",
```
Comment (lionelB) [inline packages/code-du-travail-data/dataset/fiches_service_public/package.json:10]: coquinou

File: packages/code-du-travail-data/dataset/fiches_service_public/parseReference.js [hunk 18]
```diff
@@ -1,8 +1,8 @@
 // Do we really need this one ?
 const find = require("unist-util-find");
 const queryString = require("query-string");
-const cdt = require("../code_du_travail/code-du-travail.json");
-const conventions = require("../conventions-collectives/ccn-list.json");
+const cdt = require("@socialgouv/legi-data/data/LEGITEXT000006072050.json");
```
Comment (lionelB) [inline packages/code-du-travail-data/dataset/fiches_service_public/parseReference.js]: does the parseRef should be adressed in another PR ?

File: packages/code-du-travail-data/dataset/fiches_service_public/updateTheme.js [hunk 19]
```diff
@@ -1,15 +1,34 @@
-const getThemeMapping = require("./cdtnThemeSP.js");
+const themes = require("../datafiller/themes.data.json");
+const slugify = require("../../slugify");
 
 async function updateTheme(fiches) {
-  const themeMapping = await getThemeMapping();
   fiches.forEach(fiche => {
-    fiche.themeCdtn = themeMapping[fiche.id] || [];
+    const ficheUrl = `/fiche-service-public/${slugify(fiche.title)}`;
+    const theme = themes.find(theme =>
```
Comment (lionelB) [inline packages/code-du-travail-data/dataset/fiches_service_public/updateTheme.js:7]: if we have a theme.utils we could use that :) 
```js
const theme = getTheme()
if (!theme) {
  return;
}
fiche.themeSlug = theme.slug;
fiches.breadcrumbs = getBreadcrumbs(theme)
function getBreadcrumbs({title, slug, parents}) {
  return parents.map(({title, slug}) =>({title,slug})).concat({title, slug})
}
```

File: packages/code-du-travail-api/src/server/routes/themes/index.js [hunk 20]
```diff
@@ -21,12 +35,15 @@ const router = new Router({ prefix: API_BASE_URL });
...
-    label: "Thèmes",
-    children: getChildren(null)
-  };
+router.get("/themes", async ctx => {
+  const body = getRootThemesQuery({});
+  const response = await elasticsearchClient.search({
+    index: indexThemes,
+    body
+  });
+  ctx.body = response.body;
```
Comment (lionelB) [inline packages/code-du-travail-api/src/server/routes/themes/index.js]: I wonder if we should start hiding ES from now and  unwrap data here to avoid the `results.hits.hits`

File: packages/code-du-travail-frontend/pages/theme.js [hunk 21]
```diff
@@ -21,53 +21,75 @@ const {
...
   ];
 
-  const leaf = items.map((item, index) => {
-    if (index === items.length - 1) {
+  if (!items || items.length === 0) {
+    return crumbs;
+  }
+
+  const leaves = items
+    .filter(i => !!i.title)
```
Comment (lionelB) [inline packages/code-du-travail-frontend/pages/theme.js]: we shouldn't do that here

File: packages/code-du-travail-frontend/pages/theme.js [hunk 22]
```diff
@@ -21,53 +21,75 @@ const {
...
-  const root = [
+  const crumbs = [
     <Link key="root" route="themes">
       <a title="Tous les thèmes">Thèmes</a>
     </Link>
   ];
 
-  const leaf = items.map((item, index) => {
-    if (index === items.length - 1) {
+  if (!items || items.length === 0) {
```
Comment (lionelB) [inline packages/code-du-travail-frontend/pages/theme.js]: does !items really needed ?

File: packages/code-du-travail-frontend/pages/theme.js [hunk 23]
```diff
@@ -21,53 +21,75 @@ const {
...
       return { statusCode: response.status };
     }
     const theme = await response.json();
-    return {
-      data: { theme }
-    };
+    if (theme.hits.hits.length) {
+      // show a single theme
+      if (slug) {
+        return {
```
Comment (lionelB) [inline packages/code-du-travail-frontend/pages/theme.js]: should it be done on the API side ?

File: packages/code-du-travail-frontend/pages/theme.js [hunk 24]
```diff
@@ -21,53 +21,75 @@ const {
...
+    const breadcrumbs = getBreadcrumbs(
+      (theme.parents || []).concat([
+        {
+          title: theme.title,
+          slug: theme.slug
+        }
+      ])
+    );
+    const isRootTheme = !theme.title;
+    theme.children && theme.children.length && !theme.children[0].parent;
```
Comment (lionelB) [inline packages/code-du-travail-frontend/pages/theme.js]: o_O

File: packages/code-du-travail-frontend/src/common/Answer.js [hunk 25]
```diff
@@ -28,12 +28,25 @@ const BackToResultsLink = ({ query }) => {
...
+    return [];
+  }
+  return [ROOT_CRUMB]
+    .concat(items.slice(0, -1))
+    .map(({ slug, title }) => (
+      <Link key={slug} route="themes" params={{ slug }}>
+        <a title={`Voir le theme ${title}`}>{title}</a>
+      </Link>
+    ))
+    .concat(
```
Comment (lionelB) [inline packages/code-du-travail-frontend/src/common/Answer.js]: to me the last bit should be a link too so we can navigate from an answer to its theme

**Review by lionelB — CHANGES_REQUESTED** [review]
Thanks @revolunet for this awesome works !!
having few comments  and changes

File: packages/code-du-travail-data/dataset/datafiller/package.json (hunk 3 above)
Comment (revolunet) [reply packages/code-du-travail-data/dataset/datafiller/package.json]: there is no main anymore :) just yarn start. what should we put then ?

File: packages/code-du-travail-frontend/src/home/Themes.js (hunk 4 above)
Comment (revolunet) [reply packages/code-du-travail-frontend/src/home/Themes.js]: restored :)

File: packages/code-du-travail-data/dataset/datafiller/package.json (hunk 3 above)
Comment (lionelB) [reply packages/code-du-travail-data/dataset/datafiller/package.json]: indeed we can leave it or perhaps remove it ?

File: packages/code-du-travail-data/dataset/datafiller/fetch-themes.js (hunk 12 above)
Comment (revolunet) [reply packages/code-du-travail-data/dataset/datafiller/fetch-themes.js]: yep

File: packages/code-du-travail-data/dataset/datafiller/fetch-themes.js (hunk 11 above)
Comment (revolunet) [reply packages/code-du-travail-data/dataset/datafiller/fetch-themes.js]: yep :)

File: packages/code-du-travail-data/dataset/datafiller/fix-urls.js (hunk 13 above)
Comment (revolunet) [reply packages/code-du-travail-data/dataset/datafiller/fix-urls.js]: you mean... replace promises with async/await

File: packages/code-du-travail-data/dataset/datafiller/fix-urls.js (hunk 14 above)
Comment (revolunet) [reply packages/code-du-travail-data/dataset/datafiller/fix-urls.js]: or a "prestart" script ?

File: packages/code-du-travail-data/dataset/fiches_service_public/package.json (hunk 17 above)
Comment (revolunet) [reply packages/code-du-travail-data/dataset/fiches_service_public/package.json:10]: this fix `parseReference` :)

File: packages/code-du-travail-data/dataset/fiches_service_public/parseReference.js (hunk 18 above)
Comment (revolunet) [reply packages/code-du-travail-data/dataset/fiches_service_public/parseReference.js]: can be

File: packages/code-du-travail-frontend/pages/theme.js (hunk 22 above)
Comment (revolunet) [reply packages/code-du-travail-frontend/pages/theme.js]: thats where TS would help :)

File: packages/code-du-travail-frontend/src/common/Answer.js (hunk 25 above)
Comment (revolunet) [reply packages/code-du-travail-frontend/src/common/Answer.js]: indeed

File: packages/code-du-travail-data/dataset/datafiller/fix-urls.js (hunk 13 above)
Comment (lionelB) [reply packages/code-du-travail-data/dataset/datafiller/fix-urls.js]: ok it's nit :)

File: packages/code-du-travail-data/dataset/datafiller/fix-urls.js (hunk 14 above)
Comment (lionelB) [reply packages/code-du-travail-data/dataset/datafiller/fix-urls.js]: my concern was more about the expensiveness that run requests on ES two times
but yup good idea too

File: packages/code-du-travail-frontend/pages/theme.js (hunk 22 above)
Comment (lionelB) [reply packages/code-du-travail-frontend/pages/theme.js]: PLEASE DON'T :)

Comment (revolunet) [reply packages/code-du-travail-frontend/pages/theme.js]: 😂

File: packages/code-du-travail-data/dataset/datafiller/package.json (hunk 3 above)
Comment (revolunet) [reply packages/code-du-travail-data/dataset/datafiller/package.json]: removed

File: packages/code-du-travail-data/dataset/datafiller/fix-urls.js (hunk 14 above)
Comment (revolunet) [reply packages/code-du-travail-data/dataset/datafiller/fix-urls.js]: added a prestart script

File: packages/code-du-travail-data/dataset/datafiller/elastic.js (hunk 7 above)
Comment (revolunet) [reply packages/code-du-travail-data/dataset/datafiller/elastic.js]: slug is enough atm

File: packages/code-du-travail-data/dataset/fiches_ministere_travail/report.js (hunk 15 above)
Comment (revolunet) [reply packages/code-du-travail-data/dataset/fiches_ministere_travail/report.js]: what about `breadcrumbs` ?

File: packages/code-du-travail-api/src/server/routes/themes/index.js (hunk 5 above)
Comment (revolunet) [reply packages/code-du-travail-api/src/server/routes/themes/index.js]: here we return the "refs" associated to a given theme.
the theme.refs are just urls so they're resolved "at runtime" at API level.
(that was the poorly implemented part :))
i have no better idea atm

File: packages/code-du-travail-frontend/pages/theme.js (hunk 24 above)
Comment (revolunet) [reply packages/code-du-travail-frontend/pages/theme.js]: typo :)

File: packages/code-du-travail-api/src/server/routes/themes/index.js (hunk 5 above)
Comment (lionelB) [reply packages/code-du-travail-api/src/server/routes/themes/index.js]: no prolem with that, I was refering at the format of themes data 
`({title, slug, breadcrumbs, children})` 

you could have add a documents key for exemple like it was in the endpoint themes/:slug/items

https://github.com/SocialGouv/code-du-travail-numerique/pull/1207/files#diff-a4c71ec8ea30fd77783a21a706047b2fL62-L81

File: packages/code-du-travail-frontend/pages/theme.js (hunk 23 above)
Comment (revolunet) [reply packages/code-du-travail-frontend/pages/theme.js]: much better

File: packages/code-du-travail-api/src/server/routes/themes/index.js (hunk 5 above)
Comment (revolunet) [reply packages/code-du-travail-api/src/server/routes/themes/index.js]: should be resolved

**Comment by revolunet** [comment]
thanks for the review, most comments should have been addressed.

still to be fixed : ~~extraction of "sources" package~~

File: packages/code-du-travail-api/src/server/routes/themes/index.js (hunk 2 above)
Comment (revolunet) [reply packages/code-du-travail-api/src/server/routes/themes/index.js]: done

**Comment by revolunet** [comment]
@UnbearableBear 👋 si tu as une meilleure idée pour ce fix : https://github.com/SocialGouv/code-du-travail-numerique/pull/1207/commits/876781cae002797773ed29ee3b0c589fd681bc25

l'idée est de ne pas afficher le :after quand on est dans le contexte d'une liste de résultats

Capture d’écran 2019-09-02 à 13 01 24

**Comment by UnbearableBear** [comment]
@revolunet Ca me semble être une bonne idée ! Le seule autre solution que je vois serait de target le link au niveau du parent dans les styles et d'override le :after. Mais je préfère ton option.
Mais peut-être qu'il faudrait faire de Link un composant dans le storybook

**Comment by revolunet** [comment]
Merci pour tous les retours, je prépare une PR plus propre.

Une modif à prévoir :
 - le script de correction des urls du datafiller se basera sur les données JSON des fiches et plus sur ES

**Comment by revolunet** [comment]
rebased in #1263
